### ISSUE TITLE
Chrome not reachable (on second account)

### ISSUE THREAD

**Issue by mgrimace** [klept0/MS-Rewards-Farmer#10]
Hello,

I'm experiencing the same or similar [issue](https://github.com/charlesbel/Microsoft-Rewards-Farmer/issues/363) as the main branch release that appears unresolved.

After successfully completing searches, etc. with the first account, the second account errors with the message `unknown error: cannot connect to chrome at...`. A reboot of the system doesn't change, neither does re-running the script. On re-running the script, it even starts with this second account and still results in an error. 

Console output:
```bash
2023-10-20 09:51:41,942 [INFO] patching driver executable /root/.local/share/undetected_chromedriver/undetected_chromedriver
2023-10-20 09:52:42,540 [ERROR] WebDriverException: Message: unknown error: cannot connect to chrome at 127.0.0.1:42925
from chrome not reachable
Stacktrace:
#0 0x55f408d63fb3 <unknown>
#1 0x55f408a372f6 <unknown>
#2 0x55f408a22ffa <unknown>
#3 0x55f408a6ea3c <unknown>
#4 0x55f408a662a9 <unknown>
#5 0x55f408aa9ac6 <unknown>
#6 0x55f408aa0713 <unknown>
#7 0x55f408a7318b <unknown>
#8 0x55f408a73f7e <unknown>
#9 0x55f408d298d8 <unknown>
#10 0x55f408d2d800 <unknown>
#11 0x55f408d37cfc <unknown>
#12 0x55f408d2e418 <unknown>
#13 0x55f408cfb42f <unknown>
#14 0x55f408d524e8 <unknown>
#15 0x55f408d526b4 <unknown>
#16 0x55f408d63143 <unknown>
#17 0x7f36bd094ac3 <unknown>

Traceback (most recent call last):
  File "/root/MS-Rewards-Farmer/main.py", line 31, in main
    earned_points = executeBot(currentAccount, notifier, args)
  File "/root/MS-Rewards-Farmer/main.py", line 169, in executeBot
    with Browser(mobile=False, account=currentAccount, args=args) as desktopBrowser:
  File "/root/MS-Rewards-Farmer/src/browser.py", line 41, in __init__
    self.webdriver = self.browserSetup()
  File "/root/MS-Rewards-Farmer/src/browser.py", line 78, in browserSetup
    driver = webdriver.Chrome(
  File "/usr/local/lib/python3.10/dist-packages/seleniumwire/undetected_chromedriver/webdriver.py", line 61, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/undetected_chromedriver/__init__.py", line 466, in __init__
    super(Chrome, self).__init__(
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/chromium/webdriver.py", line 56, in __init__
    super().__init__(
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/remote/webdriver.py", line 205, in __init__
    self.start_session(capabilities)
  File "/usr/local/lib/python3.10/dist-packages/undetected_chromedriver/__init__.py", line 724, in start_session
    super(selenium.webdriver.chrome.webdriver.WebDriver, self).start_session(
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/remote/webdriver.py", line 289, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/remote/webdriver.py", line 344, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.10/dist-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: cannot connect to chrome at 127.0.0.1:42925
from chrome not reachable
Stacktrace:
#0 0x55f408d63fb3 <unknown>
#1 0x55f408a372f6 <unknown>
#2 0x55f408a22ffa <unknown>
#3 0x55f408a6ea3c <unknown>
#4 0x55f408a662a9 <unknown>
#5 0x55f408aa9ac6 <unknown>
#6 0x55f408aa0713 <unknown>
#7 0x55f408a7318b <unknown>
#8 0x55f408a73f7e <unknown>
#9 0x55f408d298d8 <unknown>
#10 0x55f408d2d800 <unknown>
#11 0x55f408d37cfc <unknown>
#12 0x55f408d2e418 <unknown>
#13 0x55f408cfb42f <unknown>
#14 0x55f408d524e8 <unknown>
#15 0x55f408d526b4 <unknown>
#16 0x55f408d63143 <unknown>
#17 0x7f36bd094ac3 <unknown>
```

**Comment by klept0** [comment]
I will look into it - I personally run things in isolated folders for each account but I will do some testing

**Comment by mgrimace** [comment]
> I will look into it - I personally run things in isolated folders for each account but I will do some testing

thanks and that’s an interesting strategy - how do you do isolated folders?

**Comment by klept0** [comment]
I don’t do multiple accounts when running - each account is isolated to a folder so each "session" is a specific account. I run them at random times with their respective setups.

**Comment by mgrimace** [comment]
> I don’t do multiple accounts when running - each account is isolated to a folder so each "session" is a specific account. I run them at random times with their respective setups.

Thanks, that works for me as a work-around. Steps:

- Create folder `accounta/` and folder `accountb/`
- In each folder, `git clone` the repo
- setup account 1 .json in `accounta/MS-Rewards-Farmer`, and account 2 .json in `accountb/MS-Rewards-Farmer`
- setup cron for each folder, e.g.,:

```
10 6,8,13 * * * cd /root/accounta/MS-Rewards-Farmer && /usr/bin/python3 main.py
10 7,9,14 * * * cd /root/accountb/MS-Rewards-Farmer && /usr/bin/python3 main.py
```

I’m not sure if that’s optimal, but that seems to work with accounts separated by folder, with a cron to run A at 6:10 8:10 and 1:10, and B to run at 7:10, 9:10, and 2:10. I’m not sure how else to ‘randomize’ cron other than perhaps doing a few different entries.

**Comment by klept0** [comment]
Im not 100% a fan of the sessions folder personally so on beta branch I will be poking around with either removing it or finding a better solution. Undetected Chromedriver is pretty slow on updates and it is causing a lot of issues for people including myself.

**Comment by mgrimace** [comment]
Interesting, now I'm getting the issue even using two separate folders/instances (one account runs into the same error  as above: `2023-11-01 16:53:12,796 [ERROR] WebDriverException: Message: unknown error: cannot connect to chrome at 127.0.0.1:34741`. 

Ran the following just in case:
```
cd m/MS-Rewards-Farmer
git pull
pip install --upgrade -r requirements.txt
pip install selenium --upgrade
python3 main.py
```

**Comment by mgrimace** [comment]
New set of errors (not sure if this would be better as a separate issue):

```bash
2023-11-01 19:03:02,578 [INFO] [PUNCH CARDS] Trying to complete the Punch Cards...
2023-11-01 19:03:05,118 [ERROR] NoSuchWindowException: Message: no such window: target window already closed
from unknown error: web view not found
  (Session info: chrome=119.0.6045.105)
Stacktrace:
#0 0x55aeef7265e3 <unknown>
#1 0x55aeef3e90b7 <unknown>
#2 0x55aeef3bd1b8 <unknown>
#3 0x55aeef4636ef <unknown>
#4 0x55aeef479306 <unknown>
#5 0x55aeef45dcc3 <unknown>
#6 0x55aeef4290e4 <unknown>
#7 0x55aeef42a0ae <unknown>
#8 0x55aeef6ecce1 <unknown>
#9 0x55aeef6f0b7e <unknown>
#10 0x55aeef6da4b5 <unknown>
#11 0x55aeef6f17d6 <unknown>
#12 0x55aeef6bddbf <unknown>
#13 0x55aeef714748 <unknown>
#14 0x55aeef714917 <unknown>
#15 0x55aeef725773 <unknown>
#16 0x7fece4c94ac3 <unknown>
```

**Comment by klept0** [comment]
It appears this is an issue with undetected chromedriver - this was referenced in the original repo if I am not mistaken so take a look at this and see if this resolves your issue. 

https://github.com/charlesbel/Microsoft-Rewards-Farmer/issues/278#issuecomment-1693774447

**Comment by mgrimace** [comment]
> It appears this is an issue with undetected chromedriver - this was referenced in the original repo if I am not mistaken so take a look at this and see if this resolves your issue.
> 
> [charlesbel#278 (comment)](https://github.com/charlesbel/Microsoft-Rewards-Farmer/issues/278#issuecomment-1693774447)

Running the script with `rm -rf sessions/*/*/SingletonLock && python3 main.py` seems to be the temporary solution/workaround. However, now when I start the script I get the following error. 

```bash
2023-11-02 08:30:11,885 [INFO] [DAILY SET] Completing quiz of card 2
2023-11-02 08:30:17,247 [ERROR] WebDriverException: Message: disconnected: not connected to DevTools
  (failed to check if window was closed: disconnected: not connected to DevTools)
  (Session info: chrome=119.0.6045.105)
Stacktrace:
#0 0x5620421ad5e3 <unknown>
#1 0x562041e700b7 <unknown>
#2 0x562041e5543e <unknown>
#3 0x562041e552e3 <unknown>
#4 0x562041e727e2 <unknown>
#5 0x562041f00337 <unknown>
#6 0x562041ee4cc3 <unknown>
#7 0x562041eb00e4 <unknown>
#8 0x562041eb10ae <unknown>
#9 0x562042173ce1 <unknown>
#10 0x562042177b7e <unknown>
#11 0x5620421614b5 <unknown>
#12 0x5620421787d6 <unknown>
#13 0x562042144dbf <unknown>
#14 0x56204219b748 <unknown>
#15 0x56204219b917 <unknown>
#16 0x5620421ac773 <unknown>
#17 0x7f0ab9e94ac3 <unknown>
```

However, this does appear to be separate than the lock issue/second account, do you prefer that I open issues here in this repo or in the CharlesBel?

**Comment by mattchan8** [comment]
`rm -rf sessions && python3 main.py`
try removing the entire sessions folder before running

**Comment by mgrimace** [comment]
> `rm -rf sessions && python3 main.py` try removing the entire sessions folder before running

Thanks that did appear to work, however I ran another script in the meantime so I'll test fresh in the morning tomorrow just in case. I set up a cron for the script as `10 6,8,10 * * * cd /root/MS-Rewards-Farmer && rm -rf sessions && python3 main.py` to integrate the work-around.

**Comment by klept0** [comment]
Is this still an issue or did his fix work for you?

**Comment by mgrimace** [comment]
> Is this still an issue or did his fix work for you?

`rm -rf sessions` fixed the issue for me, and works with two accounts on main.py (no need for separate folders). I use it daily in my cron as above, and also works when I run manually. Occassionally miss a few mobile searches so I run it more than once daily. I think that's because my pi-hole is rate-limiting it but I haven't bothered investigating too much yet.

**Comment by Salindris** [comment]
Sadly, I still have this issue. I delete the sessions folder before starting and have only one account set up but it still gives me the error. What I found, though, is that in the [INFO] message it say "Created proxy listening on 127.0.0.1:45625" but the error is "cannot connect to chrome at 127.0.0.1:60503". Not knowledgeable with selenium but I guess it is trying to connect to the wrong port?

2023-11-19 15:18:39,520 [INFO] Using default request storage
2023-11-19 15:18:39,660 [INFO] Created proxy listening on 127.0.0.1:45625
2023-11-19 15:18:39,776 [INFO] Using undetected_chromedriver
2023-11-19 15:18:40,650 [INFO] patching driver executable /home/<removed>/.local/share/undetected_chromedriver/undetected_chromedriver
2023-11-19 15:19:41,493 [ERROR] WebDriverException: Message: unknown error: cannot connect to chrome at 127.0.0.1:60503
from chrome not reachable
[...]
  File "/opt/scripts/<removed>/MS-Rewards-Farmer/venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 292, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/scripts/<removed>/MS-Rewards-Farmer/venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 347, in execute
    self.error_handler.check_response(response)
  File "/opt/scripts/<removed>/MS-Rewards-Farmer/venv/lib/python3.11/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: cannot connect to chrome at 127.0.0.1:60503
from chrome not reachable

As seen in the error message I use venv to avoid version conflicts with other python projects but I don't think that that's the problem.

**Comment by SgtSobieski** [comment]
I just noticed that this error also happens to other scripts on linux which try to interact with chrome. ChatGPT advices to implement some kind of waitingfunction for the brwoser. I dont know by myself how and where to implement it, so just sharing some information...

**Comment by klept0** [comment]
Drop a link and I will try and check it out later today

**Comment by SgtSobieski** [comment]
I dont have a useful link, where it has been solved, just a reference where this error also occures occasionally (https://github.com/Memessidy/protonmail_generator). From ChatGPT its this general advice:

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# Wait for the element with the ID 'example'
element = WebDriverWait(driver, 10).until(
    EC.presence_of_element_located((By.ID, 'example'))
)

**Comment by vicrg33** [comment]
I think it is related with Chrome updates and some incompatibility with the sessions folders (by removing it you fix this issue, but you have to log in again, and if you have 2FA it is time-consuming...)

**Comment by SgtSobieski** [comment]
> I think it is related with Chrome updates and some incompatibility with the sessions folders (by removing it you fix this issue, but you have to log in again, and if you have 2FA it is time-consuming...)

Did you notice that the issue wont return afterwards? I noticed the strange behaviour about the 2fa but got the feeling that it returns after some time, but I didnt follow it for sure...

**Comment by vicrg33** [comment]
> > I think it is related with Chrome updates and some incompatibility with the sessions folders (by removing it you fix this issue, but you have to log in again, and if you have 2FA it is time-consuming...)
> 
> Did you notice that the issue wont return afterwards? I noticed the strange behaviour about the 2fa but got the feeling that it returns after some time, but I didnt follow it for sure...

I feel that it returns after a Chrome version update. The last time it stopped working my Chrome version moved from 119 to 120.

**Comment by icescoop12** [comment]
> > Is this still an issue or did his fix work for you?
> 
> `rm -rf sessions` fixed the issue for me, and works with two accounts on main.py (no need for separate folders). I use it daily in my cron as above, and also works when I run manually. Occassionally miss a few mobile searches so I run it more than once daily. I think that's because my pi-hole is rate-limiting it but I haven't bothered investigating too much yet.

for me this deletes any cookies saved needed to complete the daily quiz ,do you have this problem if your continually deleteing the sessions folder before each run?
